AOTAutograd: support mutations on buffers that happen during th bw

bdhirsh opened the pull request:
I can hold off on reviews / landing until I talk to Driss and we confirm that we need this for FP8. This PR also needs testing and probably shouldn't land until Tugsuu's input mutation handling [PR](https://github.com/pytorch/pytorch/pull/111046) goes through.

What this PR tries to solve is when you have a model that tries to mutate some nn module state (a buffer), but during the **backward**. It appears that this might be necessary for FP8's delayed scaling.

Today, AOTAutograd will just not realize if you happened to mutate any graph inputs when running the backward pass, and functionalize them away but not realize that they were input mutations. This PR tries to:

(a) detect this situation (input mutations during the backward)

(b) put `copy_()`'s in the graph to properly handle the input mutation when we can. In cases where we can't keep the copy_() in the graph, we just error loudly (I imagine that these cases will be extremely rare, but we can fix them if they ever come up).

This is mostly a prototype for now, not ready for review.

I made this example locally to test out:
```
import torch

class MutatingAutogradFn(torch.autograd.Function):

    @staticmethod
    def forward(ctx, x, buf):
        ctx.save_for_backward(buf)
        return x

    @staticmethod
    def backward(ctx, x_grad):
        buf = ctx.saved_tensors[0]
        buf.add_(x_grad)
        return x_grad * 3, None

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.buf = torch.ones(2)

    @torch._dynamo.allow_in_graph
    def backward_mutating_fn(self, x, buf):
        return MutatingAutogradFn.apply(x, buf)

    def forward(self, x):
        tmp = self.backward_mutating_fn(x, self.buf)
        return tmp + self.buf

m = Mod()

x = torch.ones(2, requires_grad=True)
out = m(x)
# After the fw, buf should not have been mutated
print(m.buf)
out.sum().backward()
# bw has run, so buf should now be mutated
print(m.buf)
print(x.grad)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112906

bdhirsh committed:
[test] AOTAutograd: support mutations on buffers that happen during th bw

[ghstack-poisoned]

ezyang commented:
Hmm, this is an interesting situation. What if I said you had to use compiled autograd in this situation?

drisspg commented:
> Hmm, this is an interesting situation. What if I said you had to use compiled autograd in this situation?

@zou3519

bdhirsh commented:
Requiring compiled autograd for anyone mutating stuff in the backward is also a valid option. Here are my tentative thoughts on pros/cons of either:

**pros** (of doing it with compiled autograd)

(1) arguably, representing "mutate a buffer during the backward" can be done more elegantly with hooks than a custom autograd.Function that no-ops in the forward. Hooks that mutate something that is not an input to the hook will **need** compiled autograd, since we need dynamo to lift the nonlocal state to an input in the backward. Compare this to the autograd.Function implementation today: Driss's FP8 code is effectively manually guaranteeing that the mutated buffer will become an input to the backward graph, by saving it for backward ([code](https://github.com/pytorch-labs/float8_experimental/blob/main/float8_experimental/float8_linear.py#L49))

(2) it (hopefully) offloads complexity from AOTAutograd onto the compiled autograd code paths, which are meant to handle stuff like "mutating non-local state in the backward".

**cons**

(1) compiled autograd is an opt-in API, so there's some user churn in erroring and requiring users to enable compiled autograd to get FP8 code working (especially if it's possible to support without compiled autograd today?)

(2) The code above is still silently incorrect today, so even if we support it in compiled autograd we'll need to make sure to error when compiled autograd isn't enabled

(3) I might be wrong, but I tried out using compiled autograd on Driss's code first and massaging it to work with a hook instead, but ran into problems ([comment](https://github.com/pytorch-labs/float8_experimental/issues/106#issuecomment-1792662768)). So I think we'll need to beef up dynamo's support for custom hooks if we want this particular hook to work today.

ezyang commented:
re con (2), I think the right way to reject this is to update the backward speculation to treat mutation as illegal as well in backwards.

re con (3), yeah, compiled autograd is not entirely baked, @voznesenskym noticed this too. But compiled autograd is an important tech so we should just invest in fixing it.

But yes, the big pro for me is (2); I am just not in the mood for reviewing complicated moving backward mutation logic around, it sounds pretty hard to actually have correct all the time.

bdhirsh committed:
Update on "[test] AOTAutograd: support mutations on buffers that happen during th bw"

I can hold off on reviews / landing until I talk to Driss and we confirm that we need this for FP8. This PR also needs testing and probably shouldn't land until Tugsuu's input mutation handling [PR](https://github.com/pytorch/pytorch/pull/111046) goes through.

What this PR tries to solve is when you have a model that tries to mutate some nn module state (a buffer), but during the **backward**. It appears that this might be necessary for FP8's delayed scaling.

Today, AOTAutograd will just not realize if you happened to mutate any graph inputs when running the backward pass, and functionalize them away but not realize that they were input mutations. This PR tries to:

(a) detect this situation (input mutations during the backward)

(b) put `copy_()`'s in the graph to properly handle the input mutation when we can. In cases where we can't keep the copy_() in the graph, we just error loudly (I imagine that these cases will be extremely rare, but we can fix them if they ever come up).

This is mostly a prototype for now, not ready for review.

I made this example locally to test out:
```
import torch

class MutatingAutogradFn(torch.autograd.Function):

    staticmethod
    def forward(ctx, x, buf):
        ctx.save_for_backward(buf)
        return x

    staticmethod
    def backward(ctx, x_grad):
        buf = ctx.saved_tensors[0]
        buf.add_(x_grad)
        return x_grad * 3, None

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.buf = torch.ones(2)

    torch._dynamo.allow_in_graph
    def backward_mutating_fn(self, x, buf):
        return MutatingAutogradFn.apply(x, buf)

    def forward(self, x):
        tmp = self.backward_mutating_fn(x, self.buf)
        return tmp + self.buf

m = Mod()

x = torch.ones(2, requires_grad=True)
out = m(x)
# After the fw, buf should not have been mutated
print(m.buf)
out.sum().backward()
# bw has run, so buf should now be mutated
print(m.buf)
print(x.grad)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "[test] AOTAutograd: support mutations on buffers that happen during th bw"

I can hold off on reviews / landing until I talk to Driss and we confirm that we need this for FP8. This PR also needs testing and probably shouldn't land until Tugsuu's input mutation handling [PR](https://github.com/pytorch/pytorch/pull/111046) goes through.

What this PR tries to solve is when you have a model that tries to mutate some nn module state (a buffer), but during the **backward**. It appears that this might be necessary for FP8's delayed scaling.

Today, AOTAutograd will just not realize if you happened to mutate any graph inputs when running the backward pass, and functionalize them away but not realize that they were input mutations. This PR tries to:

(a) detect this situation (input mutations during the backward)

(b) put `copy_()`'s in the graph to properly handle the input mutation when we can. In cases where we can't keep the copy_() in the graph, we just error loudly (I imagine that these cases will be extremely rare, but we can fix them if they ever come up).

This is mostly a prototype for now, not ready for review.

I made this example locally to test out:
```
import torch

class MutatingAutogradFn(torch.autograd.Function):

    staticmethod
    def forward(ctx, x, buf):
        ctx.save_for_backward(buf)
        return x

    staticmethod
    def backward(ctx, x_grad):
        buf = ctx.saved_tensors[0]
        buf.add_(x_grad)
        return x_grad * 3, None

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.buf = torch.ones(2)

    torch._dynamo.allow_in_graph
    def backward_mutating_fn(self, x, buf):
        return MutatingAutogradFn.apply(x, buf)

    def forward(self, x):
        tmp = self.backward_mutating_fn(x, self.buf)
        return tmp + self.buf

m = Mod()

x = torch.ones(2, requires_grad=True)
out = m(x)
# After the fw, buf should not have been mutated
print(m.buf)
out.sum().backward()
# bw has run, so buf should now be mutated
print(m.buf)
print(x.grad)
```




[ghstack-poisoned]

ezyang reviewed (approved):
discussed in VC

drisspg commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1899,6 +1899,43 @@ def functionalized_f_helper(*args):
+                # Ban metadata mutations on fw inputs during the bw
+                if not inpt_info.mutates_metadata:
+                    assert not was_metadata_updated(before, after), \
+                        "Found a graph input that had its metadata mutated in the backward. This is not supported"
```
Curious question what is an example of metadata that could be mutated in the backward?

bdhirsh replied:
Very late reply - the only example I can think of is if you have a (horrible) custom `autograd.Function`, that directly mutates one of its `grad_out` arguments in the backward. e.g.:

```
class FuncIHopeNobodyWrites(torch.autograd.Function):
    ...
    def backward(ctx, grad_x):
        grad_x.t_(1, 0)
        return grad_x * 2
        
x = torch.ones(2, 4, requires_grad=True)
out = FuncIHopeNobodyWrites.apply(x)

gradient = torch.ones(2, 4)
# the custom func will mutate gradient directly!
out.backward(gradient)
```

bdhirsh committed:
Update on "[test] AOTAutograd: support mutations on buffers that happen during th bw"

I can hold off on reviews / landing until I talk to Driss and we confirm that we need this for FP8. This PR also needs testing and probably shouldn't land until Tugsuu's input mutation handling [PR](https://github.com/pytorch/pytorch/pull/111046) goes through.

What this PR tries to solve is when you have a model that tries to mutate some nn module state (a buffer), but during the **backward**. It appears that this might be necessary for FP8's delayed scaling.

Today, AOTAutograd will just not realize if you happened to mutate any graph inputs when running the backward pass, and functionalize them away but not realize that they were input mutations. This PR tries to:

(a) detect this situation (input mutations during the backward)

(b) put `copy_()`'s in the graph to properly handle the input mutation when we can. In cases where we can't keep the copy_() in the graph, we just error loudly (I imagine that these cases will be extremely rare, but we can fix them if they ever come up).

This is mostly a prototype for now, not ready for review.

I made this example locally to test out:
```
import torch

class MutatingAutogradFn(torch.autograd.Function):

    staticmethod
    def forward(ctx, x, buf):
        ctx.save_for_backward(buf)
        return x

    staticmethod
    def backward(ctx, x_grad):
        buf = ctx.saved_tensors[0]
        buf.add_(x_grad)
        return x_grad * 3, None

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.buf = torch.ones(2)

    torch._dynamo.allow_in_graph
    def backward_mutating_fn(self, x, buf):
        return MutatingAutogradFn.apply(x, buf)

    def forward(self, x):
        tmp = self.backward_mutating_fn(x, self.buf)
        return tmp + self.buf

m = Mod()

x = torch.ones(2, requires_grad=True)
out = m(x)
# After the fw, buf should not have been mutated
print(m.buf)
out.sum().backward()
# bw has run, so buf should now be mutated
print(m.buf)
print(x.grad)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "[test] AOTAutograd: support mutations on buffers that happen during th bw"

I can hold off on reviews / landing until I talk to Driss and we confirm that we need this for FP8. This PR also needs testing and probably shouldn't land until Tugsuu's input mutation handling [PR](https://github.com/pytorch/pytorch/pull/111046) goes through.

What this PR tries to solve is when you have a model that tries to mutate some nn module state (a buffer), but during the **backward**. It appears that this might be necessary for FP8's delayed scaling.

Today, AOTAutograd will just not realize if you happened to mutate any graph inputs when running the backward pass, and functionalize them away but not realize that they were input mutations. This PR tries to:

(a) detect this situation (input mutations during the backward)

(b) put `copy_()`'s in the graph to properly handle the input mutation when we can. In cases where we can't keep the copy_() in the graph, we just error loudly (I imagine that these cases will be extremely rare, but we can fix them if they ever come up).

This is mostly a prototype for now, not ready for review.

I made this example locally to test out:
```
import torch

class MutatingAutogradFn(torch.autograd.Function):

    staticmethod
    def forward(ctx, x, buf):
        ctx.save_for_backward(buf)
        return x

    staticmethod
    def backward(ctx, x_grad):
        buf = ctx.saved_tensors[0]
        buf.add_(x_grad)
        return x_grad * 3, None

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.buf = torch.ones(2)

    torch._dynamo.allow_in_graph
    def backward_mutating_fn(self, x, buf):
        return MutatingAutogradFn.apply(x, buf)

    def forward(self, x):
        tmp = self.backward_mutating_fn(x, self.buf)
        return tmp + self.buf

m = Mod()

x = torch.ones(2, requires_grad=True)
out = m(x)
# After the fw, buf should not have been mutated
print(m.buf)
out.sum().backward()
# bw has run, so buf should now be mutated
print(m.buf)
print(x.grad)
```




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'There are lots of failure after this change https://hud.pytorch.org/pytorch/pytorch/commit/c8974d649d684a33a5c02a0b112a6e0743201d97, this is probably a landrace' -c landrace

bdhirsh commented:
closing in favor of https://github.com/pytorch/pytorch/pull/114953 (I had some trouble updating this PR with ghstack so I made a new one)
